presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.expressions.translator;
+
+import com.facebook.presto.spi.function.FunctionMetadataManager;
+import com.facebook.presto.spi.function.StandardFunctionResolution;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.DeterminismEvaluator;
+import com.facebook.presto.spi.relation.InputReferenceExpression;
+import com.facebook.presto.spi.relation.LambdaDefinitionExpression;
+import com.facebook.presto.spi.relation.LogicalRowExpressions;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionVisitor;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+public class RowExpressionTreeTranslator<T, C>
+{
+    private final RowExpressionTranslator<T, C> rowExpressionTranslator;
+    private final RowExpressionVisitor<TranslatedExpression<T>, C> visitor;
+
+    private RowExpressionTreeTranslator(RowExpressionTranslator<T, C> rowExpressionTranslator)
+    {
+        this.rowExpressionTranslator = rowExpressionTranslator;
+        this.visitor = new TranslatingVisitor();
+    }
+
+    public RowExpressionVisitor<TranslatedExpression<T>, C> getVisitor()
+    {
+        return visitor;
+    }
+
+    public static <T, C> TranslatedExpression<T> translateWith(
+            RowExpression expression,
+            DeterminismEvaluator determinismEvaluator,
+            StandardFunctionResolution functionResolution,
+            FunctionMetadataManager functionMetadataManager,
+            RowExpressionTranslator<T, C> translator,
+            C context)
+    {
+        LogicalRowExpressions logicalRowExpressions = new LogicalRowExpressions(determinismEvaluator, functionResolution, functionMetadataManager);
+        return logicalRowExpressions.convertToConjunctiveNormalForm(expression).accept(new RowExpressionTreeTranslator<>(translator).visitor, context);
+    }
+
+    private class TranslatingVisitor
+            implements RowExpressionVisitor<TranslatedExpression<T>, C>
+    {
+        @Override
+        public TranslatedExpression<T> visitCall(CallExpression call, C context)
+        {
+            return rowExpressionTranslator.translateCall(call, context, RowExpressionTreeTranslator.this);
+        }
+
+        @Override
+        public TranslatedExpression<T> visitInputReference(InputReferenceExpression reference, C context)
+        {
+            throw new UnsupportedOperationException("Cannot translate RowExpression that contains inputReferenceExpression");
```

Comment:
Add a comment: "InputReferenceExpression should only be used by Presto engine rather than connectors"